Convert dates using to_time rather than to_datetime for comparison

wjessop opened the pull request:
Fixes #36462, replicated in Rails master, Ruby 2.6.3

`to_datetime` creates an implicit +0000 UTC offset throwing comparisons with positive UTC offset times out, whereas `to_time` respects the current TZ.

### Summary

The original issue that can be summed up by this code:

```
$ TZ=Europe/London irb -r 'active_support/all'
> current_time = Time.parse('Tue, 11 Jun 2019 00:30:00 BST +01:00')
=> 2019-06-11 00:30:00 +0100
> current_date = current_time.to_date
=> Tue, 11 Jun 2019
> current_date <= current_time
=> false
```

The user is expecting the comparison to return true:

```
> current_date <= current_time
=> true
```

but it returns false.

### Other Information

Why does it do that and why is this an issue? If we expand the variables `2019-06-11 00:30:00 +0100` unzoned is effectively the same as `2019-06-10 23:30:00 +0000` (ie. UTC) for the comparison. But the date, `Tue, 11 Jun 2019`, contains no time or zone information and so when we expand this to a time, as happens in [the comparison code](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/date/calculations.rb#L1360) we get `2019-06-11 00:00:00 +0000`. The comparison is effectively therefore:

```
2019-06-11 00:00:00 +0000 <= 2019-06-10 23:30:00 +0000
```

This is clearly going to return the unexpected false result (note day of month). So is it reasonable for the user to expect the comparison to return true here? I think so. If you accept that comparing dates and times are a thing, it should be possible to compare a date with a time with a positive timezone accurately, and currently the further you get from UTC the less reliable these get, here's Rails getting weird with a 6 hour 30 time difference:

```
> current_time = Time.parse('Tue, 11 Jun 2019 06:30:00 +0700')
=> 2019-06-11 06:30:00 +0700
> current_date = current_time.to_date
=> Tue, 11 Jun 2019
> current_date < current_time
=> false
```

So why does this happen? [Here is the code that does the comparison](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/date/calculations.rb#L1360) in Rails currently:

```ruby
class Date
  …
  def compare_with_coercion(other)
    if other.is_a?(Time)
      to_datetime <=> other
    else
      compare_without_coercion(other)
    end
  end
end
```

This code converts an instance of Date to a DateTime object using `to_datetime`, then compares that to the Time object. This is the root of the issue as I hope to explain. In Ruby the Date object doesn't contain any time or zone data (which seems reasonable):

```
> Time.now
=> 2019-06-29 12:43:20 +0100
> Date.today.strftime("%z")
=> "+0000"
```

This means that when calling `to_datetime` on it there is no time or zone data to work with in the Date object itself, so a default value is used.

```
$ irb -r 'active_support/all'
> td = Time.now.to_date
=> Sat, 29 Jun 2019
> td.strftime("%z")
=> "+0000"

> td.to_datetime
=> Sat, 29 Jun 2019 00:00:00 +0000
> td.to_datetime.strftime("%z")
=> "+0000"

> td.to_time
=> 2019-06-29 00:00:00 +0100
> td.to_time.strftime("%z")
=> "+0100"
```

This difference between to_time and to_datetime seems like a bug in Ruby, though perhaps a hard one to change given the amount of code out there that will be relying on the current behaviours :)

Anyway, what this means for Rails is that the conversion of Date to DateTime creates *implicit* 00:00:00 +0000 time and zone data, which will influence the comparison between times with an offset. When we compare a time to a date using this code we are effectively moving the *date* to UTC, but leaving the *time* in it's original timezone before doing the comparison.

We can fix this though by changing the conversion in `compare_with_coercion` to `to_time`:

```
def compare_with_coercion(other)
  if other.is_a?(Time)
    to_time <=> other
  else
    compare_without_coercion(other)
  end
end
```

This works because of that subtle difference in Ruby mentioned above. `to_datetime` defaults to no zone data (+0000), but `to_time` defaults to the current TZ:

```
$ TZ=US/Eastern irb -r 'active_support/all'
> Date.today.to_time
=> 2019-06-30 00:00:00 -0400
```

Comparisons now happen as you might expect:

```
$ TZ=Europe/London irb -r 'active_support/all'
> current_time = Time.parse('Tue, 11 Jun 2019 00:30:00 BST +01:00')
=> 2019-06-11 00:30:00 +0100
> current_date = current_time.to_date
=> Tue, 11 Jun 2019
> current_date <= current_time
=> true
```

This works just great, except date.to_time doesn't actually take `Time.zone` into account which can be set using the activesupport method `Time.use_zone`, it only takes the TZ set for the process environment. To handle this condition we need to branch, using DateTime to create a comparable object, because (as far as I can tell) it's the cleanest way of creating a time like object but pinning the conversion to a specified zone

```ruby
if Time.zone
  DateTime.new(year, month, day, 0, 0, 0, Time.zone.now.formatted_offset)
else
  to_time
end
```

It looks like this code was originaly `to_time` but was changed to `to_datetime` in commit f9c2f76bea41c4a7ad1419844c6102211787c458, so if there's a technical reason for it to be `to_datetime` DHH might know.

wjessop committed:
Convert dates considering Time.zone and the process TZ env

Fixes #36462

to_datetime creates an implicit +0000 UTC offset throwing comparisons with positive UTC offset times out, whereas to_time respects the current TZ. We need to use both to ensure we do the conversion and comparison in the right context.

wjessop commented:
It occurred to me that I'm not sure how comfortable I'd feel with this getting backported if it gets accepted, even though the original barg report was for 5.2. I think that *technically* it's a bug fix and not a new feature as my reading of the current situation is that date comparison is pretty broken for affected timezones, but it does introduce some new, albeit necessary, behaviour (the comparison with the configured tz) to fix the bug.

The worry is that there may be workarounds people have in place that this code will break.

tenderlove reviewed (approved):
TBH, I think that comparing a `Date` and a `Time` doesn't make sense.  If I were to think of a date in terms of time, it would be more like a time span, and comparing a specific time to a timespan that contains that time doesn't make sense really.

It's like asking "what should `(1..6) < 2` return?".

IMO this should raise an exception, but I guess that boat has already sailed, so we need to pick a time during the date that makes sense.  Using the same TZ is obvious, and the very start of the day (0, 0, 0) seems fine too.

matthewd commented on activesupport/lib/active_support/core_ext/date/calculations.rb:
```diff
@@ -133,14 +133,26 @@ def change(options)
   def compare_with_coercion(other)
     if other.is_a?(Time)
-      to_datetime <=> other
+      convert_for_comparison_with_time <=> other
```
Is this `beginning_of_day`? If so, let's call that. If not, I think I need to more fully understand why we need to invent a new conversion.

wjessop replied:
Yup, seems to work!

matthewd commented on activesupport/lib/active_support/core_ext/date/calculations.rb:
```diff
@@ -133,14 +133,26 @@ def change(options)
+  private
+    def convert_for_comparison_with_time
+      if Time.zone
+        DateTime.new(year, month, day, 0, 0, 0, Time.zone.now.formatted_offset)
```
This part is definitely wrong: the offset at `now` and the offset at a given date may be wildly different, between DST and actual TZ changes.

wjessop commented:
Thanks for the feedback @matthewd. Have pushed a change taking it into account. Not sure the test failures are related to my code though.

wjessop commented:
As I thought, rebased master and it's green.

wjessop commented:
Any more thoughts on this @matthewd?

matthewd commented:
Sorry 🙈

I think we can cut the documentation comment back down to something similar to its previous content (maybe with the insertion of a few words like "midnight" and "current timezone"?) -- both because it's now describing behaviour that's encapsulated elsewhere, and because even when the implementation was here, it was a bit more code-descriptive than we want in the API docs.

Other than that: this sounds pretty changelog-worthy, so if you can insert an entry for it (e.g. something like "Make Date-to-Time comparisons use midnight in the current time zone"?), I'll try to get it merged before that becomes a painful rebase target.

matthewd commented:
It now occurs to me that there's an argument to be made for using the compared Time's TZ if it has one... hmmm. 🤷🏻‍♂️

wjessop commented:
I've simplified the comment and added a CHANGELOG entry, thanks for taking another look!

I feel like we have to pin the TZ of the date to something static. It seems like when the original date data was created that is where the implicit TZ is set. If I create some date data and had to ask at the point of creation "what timezone would you say this date is in", I think I'd go with the TZ the Rails process was in.

Having Dates flap around between timezones willy-nilly at the whim of any Time object that comes along later seems like it would be confusing. After all, if we use the TZ of the time we could in theory have two events that happened 48.9 hours apart (according to my calculations) compare as happening on the same date, and although I'd like to see the face of someone trying to work that one out, I don't think it would be fair :)

wjessop commented:
@matthewd Any more thoughts on this? Thanks!

wjessop commented:
@matthewd Have fixed the conflict in the changelog, would appreciate another look at this as it's still a bug.

eugeneius commented:
Pull requests can't be reopened if the branch has changed; if you reset to https://github.com/rails/rails/commit/038fd867dbb3cfbf7d93684e579237d1e98e09ea, I'll reopen this.

I think Matthew's last comment still needs a more concrete answer than "it seems like it would be confusing": can you give an example of a case where using the application's time zone would give a more sensible answer than using the zone of the time being compared?

jonathanhefner commented:
After digging into this more, I'm not sure this change should be made (yet).  I agree the current behavior is not ideal, but making this change would cause Time to behave differently than DateTime:

```ruby
# $ TZ=Europe/London irb -r 'active_support/all'
date = Date.new(2020, 6, 1)
time = Time.new(2020, 6, 1, 0, 30, 0, '+01:00')
datetime = DateTime.new(2020, 6, 1, 0, 30, 0, '+01:00')

date <=> time
# == -1

date <=> datetime
# == 1
```

I have a opened a [Ruby issue](https://bugs.ruby-lang.org/issues/16745) to fix this behavior in DateTime, though it may not be accepted due to backwards compatibility concerns.

kaspth commented:
@jonathanhefner awesome 🙏
